### PR TITLE
Change the ICE from #55223 to a hard error

### LINKED ISSUE #55287: const-eval allocation missing in dead_alloc_map
Spawned off of @oli-obk 's [example](https://github.com/rust-lang/rust/issues/55223#issuecomment-431811751) on #55223:

[play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=b23ea8c8d50f4a31a52780e27ef55fa5)

```rust
#![feature(const_let)]

union Foo<'a> {
    y: &'a (),
    long_live_the_unit: &'static (),
}

const FOO: &() = {
    let y = ();
    unsafe { Foo { y: &y }.long_live_the_unit }
};

fn main() {
}
```

### PR DESCRIPTION
**Pull request by oli-obk**
cc @SimonSapin 

r? @RalfJung 

fixes https://github.com/rust-lang/rust/issues/55287

### CODE REVIEW

File: src/test/ui/consts/dangling-alloc-id-ice-2.nll.stderr [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+LL | ||     b"CloseEvent" as &'static [u8],
+LL | || ]);
+   | || -- temporary value is freed at the end of this statement
+   | ||_|
+   | |__creates a temporary which is freed while still in use
+   |    cast requires that borrow lasts for `'static`
+   |
+   = warning: This error has been downgraded to a warning for backwards compatibility with previous releases.
+           It represents potential unsoundness in your code.
+           This warning will become a hard error in the future.
```
Comment (RalfJung) [inline src/test/ui/consts/dangling-alloc-id-ice-2.nll.stderr]: What is this about? We didn't get that for our promotion tests before, did we?

File: src/test/ui/consts/dangling-alloc-id-ice-2.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// https://github.com/rust-lang/rust/issues/55223
+
+struct Slice(&'static [&'static [u8]]);
+
+static MAP: Slice = Slice(&[
+    b"CloseEvent" as &'static [u8],
+]);
```
Comment (RalfJung) [inline src/test/ui/consts/dangling-alloc-id-ice-2.rs]: I don't even understand why anything dangles here...?

File: src/test/ui/consts/dangling-alloc-id-ice-2.nll.stderr (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/consts/dangling-alloc-id-ice-2.nll.stderr]: Oh this is NLL. Weird.

File: src/test/ui/consts/dangling-alloc-id-ice.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+#![feature(const_let)]
+
+union Foo<'a> {
+    y: &'a (),
+    long_live_the_unit: &'static (),
+}
+
+const FOO: &() = { //~ ERROR this constant cannot be used
+    let y = ();
+    unsafe { Foo { y: &y }.long_live_the_unit }
```
Comment (RalfJung) [inline src/test/ui/consts/dangling-alloc-id-ice.rs]: Same here, I don't understand why anything dangles. `y` is a local of the topmost frame, it should still be around when we intern? Or is there a `StorageDead` involved?

Comment (oli-obk) [reply src/test/ui/consts/dangling-alloc-id-ice.rs]: It's not the topmost scope, there's a block inside the topmost scope, dropping `y` at the end of it. The MIR of

```rust
const FOO: () = {
    let y = ();
    y
};
```

is

```rust
const FOO: () ={
    let mut _0: ();                      // return place
    scope 1 {
    }
    scope 2 {
        let _1: ();                      // "y" in scope 2 at src/main.rs:4:9: 4:10
    }

    bb0: {                              
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:4:9: 4:10
        _0 = _1;                         // bb0[1]: scope 1 at src/main.rs:5:5: 5:6
        StorageDead(_1);                 // bb0[2]: scope 0 at src/main.rs:6:1: 6:2
        return;                          // bb0[3]: scope 0 at src/main.rs:3:1: 6:3
    }
}
```

File: src/test/ui/consts/dangling-alloc-id-ice-2.rs (hunk 2 above)
Comment (oli-obk) [reply src/test/ui/consts/dangling-alloc-id-ice-2.rs]: it shouldn't, but it does. That's the regression. I think it's related to the explicit coercion cast

File: src/test/ui/consts/dangling-alloc-id-ice.rs (hunk 3 above)
Comment (RalfJung) [reply src/test/ui/consts/dangling-alloc-id-ice.rs]: So it is a `StorageDead`.  I think we may want to change this in the future, but for now I agree this PR is better than the status quo.

File: src/test/ui/consts/dangling-alloc-id-ice-2.rs (hunk 2 above)
Comment (RalfJung) [reply src/test/ui/consts/dangling-alloc-id-ice-2.rs]: But it's a regression not fixed by this PR, right? This test makes it look like what it tests for is intended behavior.

Comment (oli-obk) [reply src/test/ui/consts/dangling-alloc-id-ice-2.rs]: No, this PR just fixes the ICE to be a validation error. I commented on the issue and pinged the appropriate people. I don't know how to fix this regression

File: src/librustc_mir/interpret/memory.rs [hunk 4]
```diff
@@ -732,6 +732,11 @@ where
             if self.alloc_map.contains_key(&alloc) {
                 // Not yet interned, so proceed recursively
                 self.intern_static(alloc, mutability)?;
+            } else if self.dead_alloc_map.contains_key(&alloc) {
+                // danging pointer
```
Comment (estebank) [inline src/librustc_mir/interpret/memory.rs]: "danging" 👀

**Comment by RalfJung** [comment]
@bors r+

(You are playing with fire, this PR is part of a rollup which might not pick up your typo fix. Let's see.^^)

**Comment by kennytm** [comment]
@bors r-

This PR caused the rollup failure in https://github.com/rust-lang/rust/pull/55382#issuecomment-433427454

File: src/test/ui/consts/dangling-alloc-id-ice.stderr [hunk 5]
```diff
@@ -0,0 +1,13 @@
+error: this constant cannot be used
```
Comment (kennytm) [inline src/test/ui/consts/dangling-alloc-id-ice.stderr]: This error message has been changed to "any use of this value will cause an error" by #53821.

**Comment by TimNN** [comment]
Ping from triage @RalfJung: It looks like this PR is ready to be reviewed again.

**Comment by estebank** [comment]
Looking at the ICE fix commit, it is small enough that I feel confident giving my r+, assigning the approval to @RalfJung as he reviewed the other commits.

@bors r=RalfJung

**Comment by RalfJung** [comment]
Ah, sorry for that. GitHub doesn't send emails when you change labels so I had no idea this was waiting for me.

**Pull request merged**